scripts/build_kb.py

```diff
@@ -33,41 +33,62 @@ def search_knowledge_bases(root_path: str):
     elif isfile(root_path):
         with open(join(root_path), 'r') as root_file:
             for line in root_file.readlines():
+                # ignore comments and empty lines
                 line = line.replace('\n', '')
-                absolute_path = abspath(join(dirname(root_path), line))
                 # note: with current implementation, line is considered a comment if it's the first character in the line
-                #ignore paths we've already checked, comments and empty lines
-                if absolute_path in paths or line.startswith('#') or re.match(r"^\s*$", line):
+                if line.startswith('#') or re.match(r"^\s*$", line):
                     continue
+                elif line.startswith('!'):
+                    absolute_path = abspath(join(dirname(root_path), line[1:]))
+                    exclude_paths.add(absolute_path)
                 else:
-                    #recursively check each repo entry
-                    search_knowledge_bases(absolute_path)
+                    absolute_path = abspath(join(dirname(root_path), line))
+                    # ignore paths we've already checked
+                    if absolute_path not in paths:
+                        # recursively check each repo entry
+                        search_knowledge_bases(absolute_path)
                     
 
     else:
         print("Folder", root_path, "is not found.")
         exit(1)
 
+# return a file/dir name if it shouldn't be copied (returns .git folder and paths inside exclude_paths)
+def ignore_files(directory, contents):
+    ignored_files = set()
+    # for each ignored path
+    for path in exclude_paths:
+        # check if a child of this folder is inside the excluded path
+        for f in contents:
+            if abspath(join(directory, f)) == path:
+                ignored_files.add(f)
+            # ignore .git folders. Just an example of a custom ignore rule
+            if f == ".git":
+                ignored_files.add(f)
+    # return a set of filenames that should be excluded
+    print(ignored_files)
```

Comment:
Can we use logger instead of print?

Reply:
It was not intended to print it at all. I'll delete that print, thanks